File: neuroops-agent/sdk/python/neuroopssdk/logger.py
#  NeurOps Agent — Python SDK Logger
#  Copyright (c) NeurOps 2025. All rights reserved.
#
#  Structured, levelled, daily-rotating logger.
#  Writes JSON lines to logs/<component>-<date>.log AND stdout.
#
#  Usage:
#    from neuroopssdk.logger import Logger
#    log = Logger("my_plugin", level=2)
#    log.info("Plugin started")
#    log.errorf("Connection failed: %s", str(err))

import json
import os
import sys
import threading
import traceback
from datetime import datetime


class Logger:

    TRACE = 0
    DEBUG = 1
    INFO  = 2
    WARN  = 3
    ERROR = 4
    FATAL = 5

    _LEVEL_NAMES = {0: "TRACE", 1: "DEBUG", 2: "INFO", 3: "WARN", 4: "ERROR", 5: "FATAL"}

    def __init__(self, component: str, level: int = 2, log_dir: str = "logs"):
        self._component = component
        self._level     = level
        self._log_dir   = log_dir
        self._lock      = threading.Lock()
        self._file      = None
        self._date_key  = None
        os.makedirs(log_dir, exist_ok=True)
        self._rotate()

    # ── Public API ─────────────────────────────────────────────────────────

    def set_level(self, level: int):
        self._level = level

    def trace(self, msg: str):  self._log(self.TRACE, msg)
    def debug(self, msg: str):  self._log(self.DEBUG, msg)
    def info(self,  msg: str):  self._log(self.INFO,  msg)
    def warn(self,  msg: str):  self._log(self.WARN,  msg)
    def error(self, msg: str = None):
        if msg is None:
            msg = traceback.format_exc()
        self._log(self.ERROR, msg)
    def fatal(self, msg: str):  self._log(self.FATAL, msg)

    def tracef(self, fmt: str, *args): self.trace(fmt % args)
    def debugf(self, fmt: str, *args): self.debug(fmt % args)
    def infof(self,  fmt: str, *args): self.info(fmt  % args)
    def warnf(self,  fmt: str, *args): self.warn(fmt  % args)
    def errorf(self, fmt: str, *args): self.error(fmt % args)
    def fatalf(self, fmt: str, *args): self.fatal(fmt % args)

    def with_component(self, name: str) -> "Logger":
        """Return a child logger with a different component label."""
        child = Logger.__new__(Logger)
        child._component = name
        child._level     = self._level
        child._log_dir   = self._log_dir
        child._lock      = self._lock
        child._file      = self._file
        child._date_key  = self._date_key
        return child

    # ── Internal ───────────────────────────────────────────────────────────

    def _log(self, level: int, msg: str):
        if level < self._level:
            return

        now    = datetime.now()
        today  = now.strftime("%Y-%m-%d")

        with self._lock:
            if today != self._date_key:
                self._rotate(today)

            record = json.dumps({
                "timestamp": now.strftime("%Y-%m-%d %H:%M:%S.%f")[:-3],
                "level":     self._LEVEL_NAMES.get(level, "INFO"),
                "component": self._component,
                "message":   str(msg),
            })
            line = record + "\n"
            try:
                if self._file:
                    self._file.write(line)
                    self._file.flush()
            except Exception:
                pass
            print(line, end="", file=sys.stdout, flush=True)

    def _rotate(self, today: str = None):
        if today is None:
            today = datetime.now().strftime("%Y-%m-%d")
        if self._file:
            try:
                self._file.close()
            except Exception:
                pass
        safe_name = self._component.replace("/", "-").replace("\\", "-")
        path = os.path.join(self._log_dir, f"{safe_name}-{today}.log")
        try:
            self._file = open(path, "a", encoding="utf-8")
        except Exception:
            self._file = None
        self._date_key = today

File: neuroops-agent/internal/logger/logger.go
/*
 * NeurOps Agent — Logger
 * Copyright (c) NeurOps 2025. All rights reserved.
 *
 * Structured, levelled logger.  Writes JSON-formatted lines to a rotating
 * daily log file AND to stdout.  Each log record carries a timestamp,
 * level, component name, and free-form message.
 *
 * Log levels (numeric, matches agent.json):
 *   0 = TRACE  1 = DEBUG  2 = INFO  3 = WARN  4 = ERROR  5 = FATAL
 */

package logger

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"sync"
	"time"
)

// Level is the numeric log severity.
type Level int

const (
	TRACE Level = iota
	DEBUG
	INFO
	WARN
	ERROR
	FATAL
)

var levelNames = map[Level]string{
	TRACE: "TRACE",
	DEBUG: "DEBUG",
	INFO:  "INFO",
	WARN:  "WARN",
	ERROR: "ERROR",
	FATAL: "FATAL",
}

// record is the JSON shape written to disk / stdout.
type record struct {
	Timestamp string `json:"timestamp"`
	Level     string `json:"level"`
	Component string `json:"component"`
	Message   string `json:"message"`
	Caller    string `json:"caller,omitempty"`
}

// Logger is a thread-safe, levelled, rolling-file logger.
type Logger struct {
	mu        sync.Mutex
	component string
	level     Level
	writer    io.Writer
	logDir    string
	dateKey   string // "2006-01-02"
	file      *os.File
}

// New creates a Logger that writes to logs/<component>-<date>.log.
// If the logs directory does not exist it is created.
func New(component string, level Level) *Logger {
	l := &Logger{component: component, level: level, logDir: "logs"}
	_ = os.MkdirAll(l.logDir, 0o755)
	l.rotate()
	return l
}

// SetLevel changes the minimum level at runtime (hot-reload support).
func (l *Logger) SetLevel(level Level) {
	l.mu.Lock()
	defer l.mu.Unlock()
	l.level = level
}

// ── Public log methods ────────────────────────────────────────────────────────

func (l *Logger) Trace(msg string)  { l.log(TRACE, msg) }
func (l *Logger) Debug(msg string)  { l.log(DEBUG, msg) }
func (l *Logger) Info(msg string)   { l.log(INFO, msg) }
func (l *Logger) Warn(msg string)   { l.log(WARN, msg) }
func (l *Logger) Error(msg string)  { l.log(ERROR, msg) }
func (l *Logger) Fatal(msg string)  { l.log(FATAL, msg) }

// Tracef / Debugf / Infof / Warnf / Errorf / Fatalf — formatted variants.
func (l *Logger) Tracef(f string, a ...any) { l.log(TRACE, fmt.Sprintf(f, a...)) }
func (l *Logger) Debugf(f string, a ...any) { l.log(DEBUG, fmt.Sprintf(f, a...)) }
func (l *Logger) Infof(f string, a ...any)  { l.log(INFO, fmt.Sprintf(f, a...)) }
func (l *Logger) Warnf(f string, a ...any)  { l.log(WARN, fmt.Sprintf(f, a...)) }
func (l *Logger) Errorf(f string, a ...any) { l.log(ERROR, fmt.Sprintf(f, a...)) }
func (l *Logger) Fatalf(f string, a ...any) { l.log(FATAL, fmt.Sprintf(f, a...)) }

// WithComponent returns a child logger sharing the parent's file but with
// a different component label (useful for subsystems).
func (l *Logger) WithComponent(name string) *Logger {
	l.mu.Lock()
	defer l.mu.Unlock()
	return &Logger{
		component: name,
		level:     l.level,
		writer:    l.writer,
		logDir:    l.logDir,
		dateKey:   l.dateKey,
		file:      l.file,
	}
}

// ── Internal ──────────────────────────────────────────────────────────────────

func (l *Logger) log(level Level, msg string) {
	if level < l.level {
		return
	}

	now := time.Now()
	today := now.Format("2006-01-02")

	l.mu.Lock()
	defer l.mu.Unlock()

	// Roll file if day changed.
	if today != l.dateKey {
		l.rotate()
	}

	// Capture caller (skip log→log call chain: 3 frames up).
	_, file, line, ok := runtime.Caller(2)
	caller := ""
	if ok {
		caller = fmt.Sprintf("%s:%d", filepath.Base(file), line)
	}

	r := record{
		Timestamp: now.Format("2006-01-02 15:04:05.000"),
		Level:     levelNames[level],
		Component: l.component,
		Message:   msg,
		Caller:    caller,
	}

	data, _ := json.Marshal(r)
	line2 := string(data) + "\n"

	if l.writer != nil {
		_, _ = fmt.Fprint(l.writer, line2)
	}
	_, _ = fmt.Fprint(os.Stdout, line2)
}

func (l *Logger) rotate() {
	today := time.Now().Format("2006-01-02")
	if today == l.dateKey {
		return
	}
	if l.file != nil {
		_ = l.file.Close()
	}
	name := filepath.Join(l.logDir, fmt.Sprintf("%s-%s.log", strings.ReplaceAll(l.component, "/", "-"), today))
	f, err := os.OpenFile(name, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0o644)
	if err == nil {
		l.file = f
		l.writer = f
	}
	l.dateKey = today
}
